trace_decoder/src/protocol_processing.rs

```diff
@@ -0,0 +1,173 @@
+//! Logic to processes the "outer" structure of the protocol payload. All
+//! "inner" complex logic (eg. compact decoding) is handled in separate
+//! dedicated modules.
+
+use std::collections::HashMap;
+
+use enum_as_inner::EnumAsInner;
+use ethereum_types::{Address, U256};
+use mpt_trie::partial_trie::HashedPartialTrie;
+use thiserror::Error;
+
+use crate::{
+    compact::compact_prestate_processing::{
+        process_compact_prestate_debug, MptPartialTriePreImages,
+    },
+    decoding_mpt::{MptTraceParsingError, MptTraceParsingResult},
+    decoding_smt::{SmtTraceParsingError, SmtTraceParsingResult},
+    processed_block_trace::ProcessedBlockTrace,
+    processed_block_trace_mpt::{MptProcessedBlockTracePreImages, ProcessingMeta},
+    processed_block_trace_smt::SmtProcessedBlockTracePreImages,
+    trace_protocol::{
+        BlockTrace, MptBlockTraceTriePreImages, MptCombinedPreImages,
+        MptSeparateStorageTriesPreImage, MptSeparateTriePreImage, MptSeparateTriePreImages,
+        MptTrieCompact, MptTrieUncompressed, SingleSmtPreImage, SmtBlockTraceTriePreImages,
+        TriePreImage,
+    },
+    types::HashedAccountAddr,
+};
+
+const COMPATIBLE_HEADER_VERSION: u8 = 1;
+
+pub type TraceParsingResult<T> = Result<T, TraceParsingError>;
+
+// TODO: Split into two error types based on if `UnexpectedTraceFormat` can
+// occur or not...
+#[derive(Clone, Debug, Error)]
+pub enum TraceParsingError {
+    #[error(transparent)]
+    Mpt(#[from] Box<MptTraceParsingError>),
+
+    #[error(transparent)]
+    Smt(#[from] SmtTraceParsingError),
+
+    #[error("Got a trace format that was not expected!")]
```

Comment:
fwiw I'd suggest splitting these into something akin to "got smt but expected mpt" and vice versa. "unexpected" isn't terribly helpful

Reply:
Agreed, I'll change this.